async_producer.go

```diff
@@ -1114,3 +1328,43 @@ func (p *asyncProducer) abandonBrokerConnection(broker *Broker) {
 
 	delete(p.brokers, broker)
 }
+
+func (p *asyncProducer) finishTransaction(commit bool) error {
+	timedOut := waitWithTimeout(&p.inFlight, p.conf.Producer.TransactionTimeout)
+	if timedOut {
+		return errors.New("timed out while waiting for messages to be processed before committing transaction")
+	}
+
+	return p.txnmgr.finishTransaction(commit)
+}
+
+func waitWithTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait() // yes there is technically a go routine leak here when there is a timeout
```

Comment:
you can get a context.Withtimeout() based on your timeout and select on ctx.Done() to remove this go routine leak, right?